communication/throw_recovery_state_machine.h

```diff
@@ -0,0 +1,152 @@
+ * 		http://rpg.ifi.uzh.ch/docs/ICRA15_Faessler.pdf
+ *
+ ******************************************************************************/
+
+
+#ifndef THROW_RECOVERY_STATE_MACHINE_H_
+#define THROW_RECOVERY_STATE_MACHINE_H_
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include "ahrs.h"
+#include "launch_detection.h"
+#include "position_estimation.h"
+#include "remote.h"
+#include "scheduler.h"
+#include "stabilisation.h"
+#include "stabilisation_copter.h"
+#include "navigation.h"
+
+
+typedef enum {
+	STATE_IDLE = 0,
+	STATE_LAUNCH_DETECTION = 1,
+	STATE_ATTITUDE_CONTROL = 2,
+	STATE_VERTICAL_VELOCITY = 3,
+	STATE_HEIGHT_CONTROL = 4,
+	STATE_HORIZONTAL_VELOCITY = 5,
+	STATE_POSITION_LOCKING = 6
+} state_custom_t;
+
+typedef struct {
+	state_custom_t state;
+
+	bool enabled; 			///< True : Throw recovery algorithm is enabled
+	bool debug;				///< True : Debug mode activated
+	bool is_initialised;	///< True : Throw recovery state machine is ready to run
+
+	stabilisation_copter_conf_t * stabilisation_copter_conf;
+
+	remote_t * remote;
+	launch_detection_t ld;
+	imu_t * imu;
+	ahrs_t * ahrs;
+	barometer_t * baro;
+	position_estimation_t * pos_est;
+	stabilisation_copter_t * stabilisation_copter;
+	navigation_t * navigation;
+} throw_recovery_state_machine_t;
+
+
+/**
+ * \brief Initialize the state machine
+ *
+ * \param state_machine				Pointer to the state machine structure
+ * \param remote					Pointer to the remote
+ * \param imu 						Pointer to the IMU structure
+ * \param ahrs 						Pointer to the AHRS structure
+ * \param pos_est 					Pointer to the position estimation structure
+ * \param stabilisation_copter 		Pointer to the stabilisation copter structure
+ * \param navigation 	 			Pointer to the navigation structure
+ *
+ * \return	True if the init succeed, false otherwise
+ */
+bool throw_recovery_state_machine_init(	throw_recovery_state_machine_t * state_machine,
+										remote_t * remote,
+										imu_t * imu, 
+										ahrs_t * ahrs, 
+										position_estimation_t * pos_est, 
+										stabilisation_copter_t * stabilisation_copter,
+										navigation_t * navigation);
+
+
+/**
+ * \brief Update the state machine
+ *
+ * \param state_machine				Pointer to the state machine structure
+ * \param controls					Pointer to the controls to set
+ *
+ * \return	Returns the result of the task
+ */
+task_return_t throw_recovery_state_machine_update(throw_recovery_state_machine_t * state_machine, control_command_t * controls);
+
+/**
+ * \brief	Resets the state machine and launch detection
+ *
+ * \param	state_machine			The pointer to the state_machine structure
+ */
+void throw_recovery_state_machine_reset(throw_recovery_state_machine_t * state_machine);
+
+/**
+ * \brief	Returns true if the state machine should be reset
+ *
+ * \param	state_machine			The pointer to the state_machine structure
+ */
+bool throw_recovery_state_machine_should_reset(throw_recovery_state_machine_t * state_machine);
```

Comment:
It is used to make it clearer why / when we reset the state_machine : https://github.com/lis-epfl/MAVRIC/blob/throw-recovery/LEQuad/src/tasks.c#L228